packages/agents/src/mcp.ts

```diff
@@ -140,28 +300,91 @@ export abstract class McpAgent<
         const url = new URL(request.url);
         const namespace = env[binding];
 
+        // Handle SSE connections
         if (request.method === "GET" && basePattern.test(url)) {
-          const object = namespace.get(namespace.newUniqueId());
+          // Create a unique session ID for this connection
+          const sessionId = namespace.newUniqueId().toString();
+
+          // Create a Transform Stream for SSE
+          const { readable, writable } = new TransformStream();
+          const writer = writable.getWriter();
+          const encoder = new TextEncoder();
+
+          // Send the endpoint event
+          const endpointMessage = `event: endpoint\ndata: ${encodeURI(`${path}/message`)}?sessionId=${sessionId}\n\n`;
+          writer.write(encoder.encode(endpointMessage));
+
+          // Get the Durable Object
+          const id = namespace.idFromString(sessionId);
+          const doStub = namespace.get(id);
+
+          // Initialize the object
           // @ts-ignore
-          await object._init(ctx.props);
-          const response = await object.onSSE(path);
+          await doStub._init(ctx.props);
 
-          // Convert headers to a plain object
-          const headerObj: Record<string, string> = {};
-          response.headers.forEach((value, key) => {
-            headerObj[key] = value;
+          // Connect to the Durable Object via WebSocket
+          let url = new URL(request.url);
+          url.searchParams.set("sessionId", sessionId);
+          const response = await doStub.fetch(
+            new Request(url, {
+              headers: {
+                Upgrade: "websocket",
+              },
+            })
+          );
+
+          // Get the WebSocket
+          const ws = response.webSocket;
+          if (!ws) {
+            console.error("Failed to establish WebSocket connection");
+            await writer.close();
+            return;
+          }
+
+          // Accept the WebSocket
+          ws.accept();
+
+          // Handle messages from the Durable Object
+          ws.addEventListener("message", async (event) => {
+            try {
+              // Send the message as an SSE event
+              const messageText = `event: message\ndata: ${event.data}\n\n`;
```

Comment:
I'd _maybe_ consider first decoding the event data here, making sure it's JSON and has the format `{ok: true, data: <mcp-response-here> }`, rather than passing it through unchecked? I suppose it's not like a Response so you wouldn't get error messages or status codes here so maybe it's not necessary, but it always feels weird to sit in the middle of things and not inspect them?